// @iota/transaction-converter/out/transaction-converter/src/index.js
"use strict";
/** @module transaction-converter */
exports.__esModule = true;
var converter_1 = require("@iota/converter");
var pad_1 = require("@iota/pad");
var transaction_1 = require("@iota/transaction");
var errors = require("../../errors");
var guards_1 = require("../../guards");
require("../../typed-array");
var types_1 = require("../../types");
/**
 * This method takes one or more transaction objects and converts them into trytes.
 *
 * ## Related methods
 *
 * To get JSON data from the `signatureMessageFragment` field of the transaction trytes, use the [`extractJSON()`]{@link #module_extract-json.extractJSON} method.
 *
 * @method asTransactionTrytes
 *
 * @summary Converts one or more transaction objects into transaction trytes.
 *
 * @memberof module:transaction-converter
 *
 * @param {Transaction | Transaction[]} transactions - Transaction objects
 *
 * @example
 * ```js
 * let trytes = TransactionConverter.asTransactionTrytes(transactionObject);
 * ```
 *
 * @return {Trytes | Trytes[]} Transaction trytes
 *
 * @throws {errors.INVALID_TRYTES}: Make sure that the object fields in the `transactions` argument contains valid trytes (A-Z or 9).
 */
function asTransactionTrytes(transactions) {
    var txTrytes = types_1.asArray(transactions).map(function (transaction) {
        return [
            transaction.signatureMessageFragment,
            transaction.address,
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.VALUE_LENGTH)(converter_1.trytesToTrits(transaction.value))),
            pad_1.padTrytes(transaction_1.OBSOLETE_TAG_LENGTH / converter_1.TRYTE_WIDTH)(transaction.obsoleteTag),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.ISSUANCE_TIMESTAMP_LENGTH)(converter_1.trytesToTrits(transaction.timestamp))),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.CURRENT_INDEX_LENGTH)(converter_1.trytesToTrits(transaction.currentIndex))),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.LAST_INDEX_LENGTH)(converter_1.trytesToTrits(transaction.lastIndex))),
            transaction.bundle,
            transaction.trunkTransaction,
            transaction.branchTransaction,
            pad_1.padTrytes(transaction_1.OBSOLETE_TAG_LENGTH / converter_1.TRYTE_WIDTH)(transaction.tag || transaction.obsoleteTag),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.ATTACHMENT_TIMESTAMP_LENGTH)(converter_1.trytesToTrits(transaction.attachmentTimestamp))),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_LENGTH)(converter_1.trytesToTrits(transaction.attachmentTimestampLowerBound))),
            converter_1.tritsToTrytes(pad_1.padTrits(transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_LENGTH)(converter_1.trytesToTrits(transaction.attachmentTimestampUpperBound))),
            transaction.nonce,
        ].join('');
    });
    return Array.isArray(transactions) ? txTrytes : txTrytes[0];
}
exports.asTransactionTrytes = asTransactionTrytes;
/**
 * This method takes 2,673 transaction trytes and converts them into a transaction object.
 *
 * ## Related methods
 *
 * To convert more than one transaction into an object at once, use the [`asTransactionObjects()`]{@link #module_transaction-converter.asTransactionObjects} method.
 *
 * To get a transaction's trytes from the Tangle, use the [`getTrytes()`]{@link #module_core.getTrytes} method.
 *
 * @method asTransactionObject
 *
 * @summary Converts transaction trytes into a transaction object.
 *
 * @memberof module:transaction-converter
 *
 * @param {Trytes} transaction - Transaction trytes
 *
 * @example
 * ```js
 * let transactionObject = TransactionConverter.asTransactionObject(transactionTrytes);
 * ```
 *
 * @return {Transaction} transactionObject - A transaction object
 *
 * @throws {errors.INVALID_TRYTES}: Make sure that the object fields in the `transaction` argument contains valid trytes (A-Z or 9).
 */
exports.asTransactionObject = function (trytes, hash) {
    if (!guards_1.isTrytesOfExactLength(trytes, transaction_1.TRANSACTION_LENGTH / converter_1.TRYTE_WIDTH)) {
        throw new Error(errors.INVALID_TRYTES);
    }
    for (var i = 2279; i < 2295; i++) {
        if (trytes.charAt(i) !== '9') {
            throw new Error(errors.INVALID_TRYTES);
        }
    }
    var trits = converter_1.trytesToTrits(trytes);
    return {
        hash: hash || converter_1.tritsToTrytes(transaction_1.transactionHash(trits)),
        signatureMessageFragment: trytes.slice(transaction_1.SIGNATURE_OR_MESSAGE_OFFSET / converter_1.TRYTE_WIDTH, transaction_1.SIGNATURE_OR_MESSAGE_LENGTH / converter_1.TRYTE_WIDTH),
        address: trytes.slice(transaction_1.ADDRESS_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.ADDRESS_OFFSET + transaction_1.ADDRESS_LENGTH) / converter_1.TRYTE_WIDTH),
        value: converter_1.value(trits.slice(transaction_1.VALUE_OFFSET, transaction_1.VALUE_OFFSET + transaction_1.VALUE_LENGTH)),
        obsoleteTag: trytes.slice(transaction_1.OBSOLETE_TAG_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.OBSOLETE_TAG_OFFSET + transaction_1.OBSOLETE_TAG_LENGTH) / converter_1.TRYTE_WIDTH),
        timestamp: converter_1.value(trits.slice(transaction_1.ISSUANCE_TIMESTAMP_OFFSET, transaction_1.ISSUANCE_TIMESTAMP_OFFSET + transaction_1.ISSUANCE_TIMESTAMP_LENGTH)),
        currentIndex: converter_1.value(trits.slice(transaction_1.CURRENT_INDEX_OFFSET, transaction_1.CURRENT_INDEX_OFFSET + transaction_1.CURRENT_INDEX_LENGTH)),
        lastIndex: converter_1.value(trits.slice(transaction_1.LAST_INDEX_OFFSET, transaction_1.LAST_INDEX_OFFSET + transaction_1.LAST_INDEX_LENGTH)),
        bundle: trytes.slice(transaction_1.BUNDLE_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.BUNDLE_OFFSET + transaction_1.BUNDLE_LENGTH) / converter_1.TRYTE_WIDTH),
        trunkTransaction: trytes.slice(transaction_1.TRUNK_TRANSACTION_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.TRUNK_TRANSACTION_OFFSET + transaction_1.TRUNK_TRANSACTION_LENGTH) / converter_1.TRYTE_WIDTH),
        branchTransaction: trytes.slice(transaction_1.BRANCH_TRANSACTION_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.BRANCH_TRANSACTION_OFFSET + transaction_1.BRANCH_TRANSACTION_LENGTH) / converter_1.TRYTE_WIDTH),
        tag: trytes.slice(transaction_1.TAG_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.TAG_OFFSET + transaction_1.TAG_LENGTH) / converter_1.TRYTE_WIDTH),
        attachmentTimestamp: converter_1.value(trits.slice(transaction_1.ATTACHMENT_TIMESTAMP_OFFSET, transaction_1.ATTACHMENT_TIMESTAMP_OFFSET + transaction_1.ATTACHMENT_TIMESTAMP_LENGTH)),
        attachmentTimestampLowerBound: converter_1.value(trits.slice(transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_OFFSET, transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_OFFSET + transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_LENGTH)),
        attachmentTimestampUpperBound: converter_1.value(trits.slice(transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_OFFSET, transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_OFFSET + transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_LENGTH)),
        nonce: trytes.slice(transaction_1.TRANSACTION_NONCE_OFFSET / converter_1.TRYTE_WIDTH, (transaction_1.TRANSACTION_NONCE_OFFSET + transaction_1.TRANSACTION_NONCE_LENGTH) / converter_1.TRYTE_WIDTH)
    };
};
/**
 * This method takes an array of transaction hashes and returns a mapper.
 *
 * If any hashes are given, the mapper function maps them to their converted objects. Otherwise, all hashes are recalculated.
 *
 * ## Related methods
 *
 * To get a transaction's trytes from the Tangle, use the [`getTrytes()`]{@link #module_core.getTrytes} method.
 *
 * @method asTransactionObjects
 *
 * @summary Converts one or more transaction trytes into transaction objects.
 *
 * @memberof module:transaction-converter
 *
 * @param {Hash[]} [hashes] - Transaction hashes
 *
 * @example
 * ```js
 * let transactionObjectsMapper = TransactionConverter.asTransactionObjects([hashes]);
 * ```
 *
 * @return {Function} [`transactionObjectsMapper()`]{@link #module_transaction.transactionObjectsMapper}
 *
 * @throws {errors.INVALID_TRYTES}: Make sure that transcactions contains valid trytes (A-Z or 9).
 */
exports.asTransactionObjects = function (hashes) {
    /**
     * Maps the list of given hashes to a list of converted transaction objects.
     *
     * @method transactionObjectsMapper
     *
     * @param {Trytes[]} trytes - List of transaction trytes to convert
     *
     * @return {Transaction[]} List of transaction objects with hashes
     */
    return function transactionObjectsMapper(trytes) {
        return trytes.map(function (tryteString, i) { return exports.asTransactionObject(tryteString, hashes[i]); });
    };
};
exports.asFinalTransactionTrytes = function (transactions) {
    return asTransactionTrytes(transactions).slice().reverse();
};
exports.transactionObject = function (trytes) {
    /* tslint:disable-next-line:no-console */
    console.warn('`transactionObject` has been renamed to `asTransactionObject`');
    return exports.asTransactionObject(trytes);
};
exports.transactionTrytes = function (transaction) {
    /* tslint:disable-next-line:no-console */
    console.warn('`transactionTrytes` has been renamed to `asTransactionTrytes`');
    return asTransactionTrytes(transaction);
};
